Add aircleaner and humidify service to nexia climate

bdraco opened the pull request:
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

* These were removed from the original merge to reduce review scope

* Additional tests for binary_sensor, sensor, and climate states

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/12467

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Add aircleaner and humidify service to nexia climate

* These were removed from the original merge to reduce review scope

* Additional tests for binary_sensor, sensor, and climate states

frenck commented:
Instead of figuring out which entity id's to call the service on, one could use an entity service. See:

<https://developers.home-assistant.io/docs/dev_101_services#entity-services>

bdraco commented:
> Instead of figuring out which entity id's to call the service on, one could use an entity service. See:
> 
> https://developers.home-assistant.io/docs/dev_101_services#entity-services
@frenck  Thanks for looking at this!

I didn't use async_register_entity_service because it can't handle all the config entries. 
https://github.com/home-assistant/core/pull/32826#discussion_r393456831

The goal was to be able to put all the air cleaners in "allergy mode" with a single call.

This use case is probably very specific to my needs and may not represent the user base as a whole. Switching to `async_register_entity_service` would reduce the code here quite a bit and I could still accomplish the same thing with some scripts so it doesn't make much different to me personally. 

How should I proceed?

frenck commented:
If I understand it correctly, basically you are trying to figure out, upfront, which entities apply. Which is what causes the need to do this right now, instead, can't the entity itself figure out if it should apply the requested setting?

That way, the entity ID's in the service call, will all receive the requested setting and decide themself apply it our not (by determining e.g., zone or capabilities).

Which would allow for use of the entity services. The entity service do more than just entity selection (they also apply permission and context). So I think that should be the preferred method.

bdraco commented:
Thanks for the additional color, I guess I need to setup areas :facepalm:.  I’ll do some testing with this after some sleep.

MartinHjelmare commented:
I think there's currently a flaw in `async_register_entity_service` where it overwrites the service handler for the platform name if more then one platform of the same platform domain and integration domain registers services.

bdraco commented:
@frenck @MartinHjelmare Given the flaw in `async_register_entity_service` and the no storage policy, I'm going to rework the existing code to send a signal keyed on the thermostat id if that sounds ok?

MartinHjelmare commented:
Sounds ok to me.

bdraco commented:
> Sounds ok to me.

I don't think this approach is going to work either as we don't have a way to target the thermostats the zones are on. With 16 zones and 4 thermostats thats 16 api call when only 4 are needed. 

I'm going to have to restrict it to per entity only and rework once `async_register_entity_service` is fixed

bdraco committed:
Switch to signals for services

Get rid of everywhere we call device and change to zone or thermostat
as it was too confusing

Renames to make it clear that zone and thermostat are tightly coupled

bdraco committed:
Make scene activation responsive

bdraco commented:
Should be ready for another review pass.

balloob commented on homeassistant/components/nexia/climate.py:
```diff
@@ -318,118 +314,163 @@ def set_temperature(self, **kwargs):
-        }
+        data = super().device_state_attributes
+
+        data.update({ATTR_ZONE_STATUS: self._zone.get_status()})
```
This will be faster.

```suggestion
        data[ATTR_ZONE_STATUS] = self._zone.get_status()
```

bdraco replied:
Thanks. Much cleaner.

bdraco committed:
no need to use update for only one key/value

bdraco committed:
stray comma

balloob commented on homeassistant/components/nexia/scene.py:
```diff
@@ -36,33 +32,25 @@ class NexiaAutomationScene(NexiaEntity, Scene):
         """Activate an automation scene."""
-        self._automation.activate()
+        await self.hass.async_add_executor_job(self._automation.activate)
+        await asyncio.sleep(SCENE_ACTIVATION_TIME)
```
Shouldn't we just schedule an async refresh to be done after `SCENE_ACTIVATION_TIME` ? The command has already been issued, it's just the state update causing us to sleep. That means that any script/automation will now also have to wait for the state change, that seems unnecesssary.

bdraco replied:
Yes, sleep is the solution of last resort :)

bdraco replied:
switched to async_call_later

bdraco replied:
That didn't work
```
Exception in callback async_track_point_in_utc_time.<locals>.point_in_time_listener(<Event time_c....003775-05:00>) at /usr/src/homeassistant/homeassistant/helpers/event.py:234
```
I screwed something up.. rechecking

bdraco replied:
fixed

balloob commented on homeassistant/components/nexia/services.py:
```diff
@@ -0,0 +1,58 @@
+    )
+
+    async def _aircleaner_set_service(service):
+        async_dispatcher_send(hass, SIGNAL_AIRCLEANER_SERVICE, service.data)
```
Calling a service should wait until the request is done. By sending a dispatcher, the service task will return right away before the service call is actually executed.

bdraco replied:
@balloob I'm kinda stuck with this approach due to https://github.com/home-assistant/core/pull/33078#issuecomment-602007549

balloob replied:
Oh shit yes. Let me see if I can whip something up.

bdraco replied:
Converted :)

balloob reviewed (approved):
Ok to merge when comments addressed.

bdraco committed:
use async_call_later

bdraco committed:
its async, need ()s

balloob commented on homeassistant/components/nexia/scene.py:
```diff
@@ -52,5 +52,5 @@ def icon(self):
         async_call_later(
-            self.hass, SCENE_ACTIVATION_TIME, self._coordinator.async_refresh
+            self.hass, SCENE_ACTIVATION_TIME, self._coordinator.async_refresh()
         )
```
the callback of `async_call_later` is passed the time it is being called. By passing a `coro` you bypass this, but now if Home Assistant were to quit before this was called, it will print a warning for having an unawaited coro. 

It's cleaner if you create a function to handle it.

```suggestion
        async def refresh_callback(_):
            await self._coordinator.async_refresh()
        
        async_call_later(
            self.hass, SCENE_ACTIVATION_TIME, refresh_callback
        )
```

bdraco replied:
Thanks for the detailed explanation. Very helpful. All fixed up

bdraco committed:
cleaner

bdraco commented:
Testing https://github.com/home-assistant/core/pull/33176 now

bdraco committed:
merge entity platform services testing branch

bdraco commented:
Retested again, should be good to go if CI says ok
